src/tools/rbd/action/Create.cc

```diff
@@ -25,7 +27,191 @@ void get_arguments(po::options_description *positional,
   at::add_size_option(options);
+  at::add_no_progress_option(options);
+}
+
+void thick_provision_writer_completion(rbd_completion_t, void *);
+
+struct thick_provision_writer {
+  librbd::Image *image;
+  Mutex lock;
+  Cond cond;
+  bufferlist bl;
+  uint64_t chunk_size;
+  const int block_size;
+  uint64_t concurr;
+  struct {
+    uint64_t in_flight;
+    int io_error;
+  } io_status;
+
+  // Constructor
+  explicit thick_provision_writer(librbd::Image *i, librbd::ImageOptions &o)
+    : image(i),
+      lock("thick_provision_writer::lock"),
+      block_size(512) // 512 Bytes
+  {
+    // If error cases occur, the code is aborted, because
+    // constructor cannot return error value.
+    assert(g_conf != nullptr);
+    bl.append_zero(block_size);
+
+    librbd::image_info_t info;
+    int r = image->stat(info, sizeof(info));
+    assert(r >= 0);
+    uint64_t order;
+    if (info.order == 0) {
+      order = g_conf->get_val<int64_t>("rbd_default_order");
+    } else {
+      order = info.order;
+    }
+    chunk_size = (1ull << order);
+    if (image->get_stripe_unit() < chunk_size) {
+      chunk_size = image->get_stripe_unit();
+    }
+
+    concurr = g_conf->get_val<int64_t>("rbd_concurrent_management_ops");
+    io_status.in_flight = 0;
+    io_status.io_error = 0;
+  }
+
+  int start_io(uint64_t write_offset)
+  {
+    {
+      Mutex::Locker l(lock);
+      io_status.in_flight++;
+      if (io_status.in_flight > concurr) {
+        io_status.in_flight--;
+        return -EINVAL;
+      }
+    }
+
+    librbd::RBD::AioCompletion *c;
+    c = new librbd::RBD::AioCompletion(this, thick_provision_writer_completion);
+    int r;
+    r = image->aio_writesame(write_offset, chunk_size, bl, c, LIBRADOS_OP_FLAG_FADVISE_SEQUENTIAL);
+    if (r < 0) {
+      Mutex::Locker l(lock);
+      io_status.io_error = r;
+    }
+    return r;
+  }
+
+  int wait_for(uint64_t max) {
+    Mutex::Locker l(lock);
+    int r = io_status.io_error;
+
+    while (io_status.in_flight > max) {
+      utime_t dur;
+      dur.set_from_double(.2);
+      cond.WaitInterval(lock, dur);
+    }
+    return r;
+  }
+};
+
+void thick_provision_writer_completion(rbd_completion_t rc, void *pc) {
+  librbd::RBD::AioCompletion *ac = (librbd::RBD::AioCompletion *)rc;
+  thick_provision_writer *tc = static_cast<thick_provision_writer *>(pc);
+
+  int r = ac->get_return_value();
+  tc->lock.Lock();
+  if (r < 0 &&  tc->io_status.io_error >= 0) {
+    tc->io_status.io_error = r;
+  }
+  tc->io_status.in_flight--;
+  tc->cond.Signal();
+  tc->lock.Unlock();
+  ac->release();
+}
+
+int write_data(librbd::Image &image, librbd::ImageOptions &opts,
+               bool no_progress) {
+  uint64_t image_size;
+  int r = 0;
+  utils::ProgressContext pc("Thick provisioning", no_progress);
+
+  if (image.size(&image_size) != 0) {
+    return -EINVAL;
+  }
+
+  thick_provision_writer tpw(&image, opts);
+  vector<uint64_t> thread_offset;
```

Comment:
I still think you can simplify this by removing the thread offsets. You just need to sequentially move the offset through the image by ```tpw.chunk_size``` bits.